### ISSUE TITLE
[Feature] Add APM for Serverless PHP

### ISSUE THREAD

**Issue by eddmann** [DataDog/dd-trace-php#1207]
Over the past couple of weeks we have been setting up DataDog APM within our stack.
We have found great value in enabling this on our conventional server-based PHP instances, using the DataDog Agent to forward on the tracing to DataDog.

However, over the past year we have been adopting Serverless (via AWS Lambda) - successfully running PHP applications brought in large part by the [BrefPHP](https://bref.sh/) project.
We noticed that Serverless support is present for languages such as Python/Ruby but sadly not for PHP.
We really liked the idea of getting the same level of insight into our Serverless environment as we did the conventional server-based approach so we set aside sometime to explore if this would be possible.

### Proof-of-concept

We were able to create a PoC based on the existing PHP-APM extension along with a couple of small modifications.
We have made this available at https://github.com/mybuilder/datadog-php-tracer-lambda-layers.

At a technical level you can see that based on how the Tracer has been designed we only need to supply a different [Transport](https://github.com/mybuilder/datadog-php-tracer-lambda-layers/blob/main/php/DDTrace/Transport/StdOutJsonStream.php) which writes these traces out to `stdout` (CloudWatch Logs) - like what is done in other languages such as Python.
We are already using this to supply custom-metrics from our Serverless application using the JSON-based format which is processed by the DataDog Forwarder.

As it stands this works surprisingly well, with only a couple of caveats:

- We cater for dropping traces within the Transport as opposed to always sending onto the DataDog Forwarder so as to not incur extra cost - we have not invested the time to explore how Python etc. does this, or if they simply send all traces along to DataDog for them to be discarded and metrics to be counted.
- We had to modify the format in which [identifiers were sent](https://github.com/mybuilder/datadog-php-tracer-lambda-layers/blob/main/php/DDTrace/Transport/StdOutJsonStream.php#L51-L67) (strings instead of integer) as the DataDog endpoint did not like this.
- We found we needed to [chunk up](https://github.com/mybuilder/datadog-php-tracer-lambda-layers/blob/main/php/DDTrace/Transport/StdOutJsonStream.php#L39-L42) the traces being sent out on the `stdout` for CloudWatch as we seemed to hit a max-line limit.
- We are having issues with distributed traces crossing the Serverless/Server-based divide, Serverless-to-Serverless traces are working as expected but not between. We have not been able to figure out what this issue is due to.

### Conclusion

At a high-level we are wondering if this is something you would look to support in the future?

We understand that PHP is not a supported runtime for AWS so providing a layer etc. maybe a little non-trivial, but providing a means to supply a `stdout` based trace output like we have shown would be great!
In regards to this there already is a [Stream](https://github.com/DataDog/dd-trace-php/blob/master/src/DDTrace/Transport/Stream.php) transport and [Json](https://github.com/DataDog/dd-trace-php/blob/master/src/DDTrace/Encoders/Json.php) encoder present in the codebase, so we are not sure if this was an intention at some point already?

**Comment by bwoebi** [comment]
Yes, we would be open to such a MR. However, such an implementation would have to be at the extension layer by now.

That implementation would need to hook into https://github.com/DataDog/dd-trace-php/blob/38c9ca6765cb6430e8bcee802aa539f7994ddf4b/ext/php8/auto_flush.c#L9.

Note that we are also open to providing an API, such that one can provide a callback from PHP code to be invoked upon flushing here.

**Comment by deleugpn** [comment]
Just wanted to share this here in case someone else in the future finds it useful:

https://blog.deleu.dev/datadog-aws-lambda-php/

TL;DR it's possible to use Datadog on Lambda without changing anything on the PHP Extension

**Comment by dzavalkin-scayle** [comment]
We managed to get it working in a slightly different way. We run Datadog agent as a Fargate task and install only `dd-tracer-php` PHP extension inside lambda and configured it to connect to external Datadog agent. This way we avoid any possible performance hits caused by running Datadog agent inside lambda. This setup is not in production yet, we are still testing but no problems were noticed so far - `dd-tracer-php` sends all spans to external Datadog agent and their appear in Datadog UI as expected.

**Comment by Anixed** [comment]
With my team we need this feature for serverless PHP.

@dzavalkin-scayle How did with this configuration in production?

**Comment by dzavalkin-scayle** [comment]
Setup I described above is for **APM Datadog integration**.
Datadog is running as an ECS/Fargate task/service and there is an NLB pointed to it to port `8126`.
On Lambda side we set the following env variables:

```env
DD_AUTOFINISH_SPANS: true
DD_ENV: envname
DD_SERVICE: servicename
DD_SERVICE_MAPPING: pdo:servicename-envname-rds,phpredis:servicename-envname-ec,redis:servicename-envname-ec,guzzle:servicename-envname-guzzle,curl:servicename-envname-curl
DD_TAGS: service:servicename
DD_TRACE_AGENT_URL: http://servicename-datadog-nlb-int-1234567890123456.elb.eu-west-1.amazonaws.com:8126
DD_TRACE_AUTO_FLUSH_ENABLED: true
DD_TRACE_CLI_ENABLED: true
DD_TRACE_CURL_ENABLED: true
DD_TRACE_ELOQUENT_ENABLED: true
DD_TRACE_ENABLED: true
DD_TRACE_GENERATE_ROOT_SPAN: true
DD_TRACE_GUZZLE_ENABLED: true
DD_TRACE_LARAVEL_ENABLED: true
DD_TRACE_PDO_ENABLED: true
```

To forward **Lambda logs** (and also logs from ECS/Fargate tasks) we use **Datadog Forwarder** (i.e. CloudWatch Logs -> DD forwarding): https://docs.datadoghq.com/logs/guide/forwarder/

**Comment by koullislp** [comment]
Would be great to have this supported by Datadog!

**Comment by astuyve** [comment]
Hey folks! Quick poll on runtimes in Lambda here:
🎉 - Using Bref FPM
🌮 - Using Bref Functions
🚀 - Other (please comment/specify)

Thanks!

**Comment by natepage** [comment]
Hello here! 

I've been playing with PHP on Serverless Aws Lambda lately, and similar to the original comment of this issue,
We are using Datadog solutions to monitor our server based applications and we would love to have the same quality of life in our serverless applications 😄 

After reading all comments here, I feel like things have improved significantly since 2023, especially with the shiny next-gen AWS Lambda Extension!

@astuyve below is our runtime setup 😄 

We are currently running the following setup:
- Symfony applications (Api and Queue Workers) using Bref Functions runtime
- Installed this PHP extension using the AWS Lambda layer provided by [Bref](https://github.com/brefphp/extra-php-extensions/blob/master/docs/datadog.md) `arn:aws:lambda:${AWS::Region}:403367587399:layer:datadog-php-82:7`
- Installed the Datadog AWS Lambda Extension, again using the AWS Lambda Layer `arn:aws:lambda:${AWS::Region}:464622532012:layer:Datadog-Extension:70`

And the experience was honestly quite smooth, installed eveyrthing and the PHP application is still running, nothing fundamentaly broken 😄 
However, going to Datadog console, I was surprised to discover that my function didn't have any traces and/or logs.

After enabling all the debug options possible, reading docs, reading code I didn't understand at first (my C and Rust aren't great haha), it looks like we are very close to a complete working solution.
We are currently facing 2 issues:

- First one is related to using the latest next-gen AWS Lambda Extension, and our Datadog organisation being PCI compliant, we are not enable to send logs, but I already created an [issue](https://github.com/DataDog/datadog-lambda-extension/issues/580) (earlier today actually) in the extension repo for that so let's not focus on that here 😄 
- The second issue is that ddtrace is failing to send the traces to the trace agent running within the lambda (started by the Lambda extension) because the HTTP request is failing validation from the trace agent. Let me explain more.

The next-gen trace agent running within the lambda function expects traces requests to contain the `Content-Length` HTTP header in order to process traces, however my understanding is because of the following logic within ddtrace coms enforcing `Transfer-Encoding: chunked` it results in curl not automatically setting the content-length header resulting in the trace agent to reject the incoming requests.

https://github.com/DataDog/dd-trace-php/blob/0051dbea15c989702e4227e5a9dbcd57d962fdac/ext/coms.c#L808

Here is an actual curl log coming from one of our applications, and we can see that the `Content-Length` header is missing:
```
  | 2025-02-28T06:14:26.707Z | * Trying 127.0.0.1:8126...
  | 2025-02-28T06:14:26.717Z | * Connected to 127.0.0.1 (127.0.0.1) port 8126
  | 2025-02-28T06:14:26.717Z | * using HTTP/1.x
  | 2025-02-28T06:14:26.718Z | > PUT /v0.4/traces HTTP/1.1
  | 2025-02-28T06:14:26.718Z | Host: 127.0.0.1:8126
  | 2025-02-28T06:14:26.718Z | Accept: */*
  | 2025-02-28T06:14:26.718Z | Datadog-Meta-Lang: php
  | 2025-02-28T06:14:26.718Z | Datadog-Meta-Lang-Interpreter: cli
  | 2025-02-28T06:14:26.718Z | Datadog-Meta-Lang-Version: 8.2.27
  | 2025-02-28T06:14:26.718Z | Datadog-Meta-Tracer-Version: 1.5.1
  | 2025-02-28T06:14:26.718Z | Transfer-Encoding: chunked
  | 2025-02-28T06:14:26.718Z | Content-Type: application/msgpack
  | 2025-02-28T06:14:26.718Z | X-Datadog-Trace-Count: 4
  | 2025-02-28T06:14:26.718Z | DD_EXTENSION \| ERROR \| Error processing traces: Missing Content-Length header
  | 2025-02-28T06:14:26.718Z | < HTTP/1.1 411 Length Required
  | 2025-02-28T06:14:26.718Z | < content-length: 68
  | 2025-02-28T06:14:26.718Z | < date: Fri, 28 Feb 2025 06:14:26 GMT
  | 2025-02-28T06:14:26.718Z | * HTTP error before end of send, stop sending
  | 2025-02-28T06:14:26.718Z | * abort upload after having sent 3479 bytes
  | 2025-02-28T06:14:26.718Z | <
  | 2025-02-28T06:14:26.718Z | * shutting down connection #0
```

It appears that if we manage to fix this communication between ddtrace and the trace agent, then APM traces should flow through, but I might be missing something here..

Sorry for the long comment, and thank you for reading!

**Comment by bwoebi** [comment]
Hey @natepage,

the new sidecar sender will send the Content-Length header. The old background sender (which is what you were looking at) does not.
To enable the sidecar set `DD_TRACE_SIDECAR_TRACE_SENDER=1` (it's enabled by default on PHP 8.3 and PHP 8.4; we're going to slowly roll it out to more customers as the number of reported issues decreases).

I hope that helps :-)

**Comment by natepage** [comment]
Hi @bwoebi, 

Thank you for you reply, 

I've done a quick update to PHP8.3, it's definitely different, but still getting errors 😄 

```
[ddtrace] [error] Failed sending traces to the sidecar: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }
[ddtrace] [datadog_sidecar::agent_remote_config] Tried to open path "/ddcfg-993-7303405563243401017" for agent remote config, but failed: Os { code: 38, kind: Unsupported, message: "Function not implemented" }
```

And also some trace logs from the Lambda extension

```
DD_EXTENSION | TRACE | File descriptor use data not found in path /proc/16/fd with pid 16
```

Do I need to update to the latest ddtrace version? I've checked and the current Bref layer is using ddtrace 1.5.1

Thank you, still feels like progress 😄

**Comment by natepage** [comment]
@bwoebi thinking about your last comment and this sidecare feature, do I understand correctly that traces will be given to the agent by creating files in a shared filesystem?

If yes, then would the AWS Lambda extension support that? If we manage to create files under `/tmp` in Lambda, as far as I know the mini agent from the AWS Lambda extension expects traces via http requests to `/<version>/traces` endpoint, unless I'm missing something 😄

**Comment by bwoebi** [comment]
No, the sidecar (separate process spawned by the ddtrace extension) will receive the data from shared memory, then send it over to the (mini) agent.

**Comment by natepage** [comment]
Hi there!

It's been a few weeks with no communication here, but things have happened so I thought it would be useful to share the exciting news here!

First I would like to start by thanking both @bwoebi and @astuyve for their help making it possible, through their guidance and changes implemented in Datadog libraries!

APM for Serverless PHP is here!!! 😄 

I'm being very cautious when saying this as I fully appreciate that the solution we've put in place requires using the latest versions of dd-trace-php/AWS Lambda Extension/etc. and not all projects would have the chance to use those shiny new things (we have a few legacy projects ourselves...).
However, if your project is fortunate enough to be running on a fairly recent version of PHP and you have the ability to upgrade your stack, the latest changes implemented over the past few weeks make monitoring your Serverless PHP application very easy!

 I thought I would share some info here about what changes made this possible, and an overview of our setup so it could potentially help others.

Here are the minimum versions required to make this setup work:

- [dd-trace-php v1.8.0](https://github.com/DataDog/dd-trace-php/releases/tag/1.8.0): enables sidecar traces sender by default in AWS Lambda and addresses readonly filesystem issues
- [datadog-lambda-extension v76](https://github.com/DataDog/datadog-lambda-extension/releases/tag/v76): fixes JSON logs in AWS Lambda

Here is an overview of our setup:

- Symfony application with PHP8.3 (Not that it really matters here but just for completeness)
- Bref PHP packages to help run in AWS Lambda (bref, symfony bridge, symfony messenger)
- Bref Function Runtime for PHP8.3
- Datadog Lambda Extension NextGen (v76) for Serverless monitoring + Logs collection
- dd-trace-php ^1.8 for Traces collection
- AWS SAM to deploy to AWS
- Docker to build the image we use within AWS Lambda
- That's it! 😄 

**Why Docker?**

We would have loved to be keep Docker out of the picture, and use Lambda layers to import the various libraries we need to run PHP, instrument it, etc. But because layers still count towards the "package size limit" imposed by AWS Lambda, we very quickly went over the 250MB allowed.
This was an acceptable tradeoff for us as we very much value having a standalone Lambda function that offers the same functionality and monitoring level as our overally complex ECS setup with application containers + datadog agent container + sidecar logs processing container...


**To summarise the solution:**

- Write you PHP application as you please (Symfony/Laravel/Vanilla/etc)
- Install and configure Bref PHP to make it work in AWS Lambda
- Create a docker image with your code, dd-trace-php, and the aws-lambda-extension
- Configure your deployment framework (Serverless/AWS SAM) to setup Datadog (config files/env vars)
- Deploy it!
